pkg/kubelet/kuberuntime/kuberuntime_gc.go

```diff
@@ -354,12 +356,25 @@ func (cgc *containerGC) evictPodLogsDirectories(allSourcesReady bool) error {
 	// Remove dead container log symlinks.
 	// TODO(random-liu): Remove this after cluster logging supports CRI container log path.
 	logSymlinks, _ := osInterface.Glob(filepath.Join(legacyContainerLogsDir, fmt.Sprintf("*.%s", legacyLogSuffix)))
+	now := time.Now()
 	for _, logSymlink := range logSymlinks {
 		if _, err := osInterface.Stat(logSymlink); os.IsNotExist(err) {
-			err := osInterface.Remove(logSymlink)
-			if err != nil {
-				klog.Errorf("Failed to remove container log dead symlink %q: %v", logSymlink, err)
+			if firstSeen, ok := cgc.symlinkFirstObservedMissing[logSymlink]; ok {
+				if now.Sub(firstSeen) > 30*time.Second {
```

Comment:
If I'm reading this correctly, we are trying to resolve the race condition by adding a wait for X seconds... tbh that strategy makes me a little nervous. In particular, it feels very difficult to effectively write tests for.

Is there instead some expected state of the world that we could wait for before we remove the `logSymlink`? I.e. a state of the world that would allow us to be confident that we avoided the race condition.

Reply:
The current formation does not introduce locking between pkg/kubelet/kuberuntime and pkg/kubelet/logs .
I think this is extensible compared to the locking approach.

Here is how log rotation works (see containerLogManager#rotateLatestLog):

1. rename current log to rotated log file (fmt.Sprintf("%s.%s", log, timestamp))
2. reopen container log
3. if #2 fails, rename rotated log file back

There is small but indeterministic amount of time during which log file doesn't exist.
This was why I chose the current approach.

Reply:
Currently logSymlinks is empty for tests from kuberuntime_gc_test :
```
	logSymlinks, _ := osInterface.Glob(filepath.Join(legacyContainerLogsDir, fmt.Sprintf("*.%s", legacyLogSuffix)))
```
We can make legacyContainerLogsDir a var so that the test can override.
The GC interval should also be made configurable (for testing).

For the new test, a goroutine would constantly rotate log file (with GC active). We can assert that the symlink to the log file is never gone.

@mattjmcnaughton 
What do you think ?

Reply:
@mattjmcnaughton @rphillips 
I have updated the PR with a new unit test.

If you think using mock is the way to go, I will refine the test.

Thanks

Reply:
I'd love to get the thoughts of another one of the folks tagged here - @Random-Liu @dims @derekwaynecarr . This diff has moved outside my area of expertise, so I'm not sure I have much to offer beyond my initial comment :)